pyadjoint/checkpointing.py

```diff
@@ -0,0 +1,305 @@
+            The current timestep.
+
+        Raises
+        ------
+        CheckpointError
+            If the checkpoint action is not supported.
+        """
+        raise CheckpointError(f"Unable to process {cp_action} while taping.")
+
+    @process_taping.register(Forward)
+    def _(self, cp_action, timestep):
+        if timestep < (cp_action.n0):
+            raise CheckpointError(
+                "Timestep is before start of Forward action."
+            )
+
+        self.tape._eagerly_checkpoint_outputs = cp_action.write_adj_deps
+
+        if timestep in cp_action:
+            self.tape.get_blocks().append_step()
+            if cp_action.write_ics:
+                # checkpoint_schedules has the forward action given by
+                # Forward(n0, n1, with_ics, with_adj_deps).
+                # If with_ics is True, then the restart forward data is
+                # checkpointed. This restart data is the data required
+                # for the forward model from the step `n0`.
+                if timestep == cp_action.n0:
+                    self.tape.latest_checkpoint = timestep
+                    self.tape.timesteps[cp_action.n0].checkpoint()
+                    self._configuration_step = timestep
+
+                # If with_adj_deps is False, then the adjoint dependencies are
+                # not required to be checkpointed.
+                if not cp_action.write_adj_deps:
+                    for step in range(self._configuration_step, timestep):
+                        for block in self.tape.timesteps[step]:
+                            for output in block.get_outputs():
+                                output._checkpoint = None
+            return True
+        else:
+            return False
+
+    @process_taping.register(EndForward)
+    def _(self, cp_action, timestep):
+        # The correct number of forward steps has been taken
+        assert timestep == self.timesteps
+        self.mode = Mode.EVALUATED
+        return True
+
+    def recompute(self, functional=None):
+        """Recompute the forward model."""
+        self.mode = Mode.RECOMPUTE
+
+        with self.tape.progress_bar("Evaluating Functional",
+                                    max=self.fwd_schedule.steps) as bar:
+            for cp_action in self.fwd_schedule:
+                self.process_operation(cp_action, bar, functional=functional)
+
+    def evaluate_adj(self, last_block, markings):
+        """Evaluate the adjoint model.
+
+        Parameters
+        ----------
+        last_block : int
+            The last block to be evaluated.
+        markings : bool
+            If True, then each block_variable of the current block will have
+            set `marked_in_path` attribute indicating whether their adjoint
+            components are relevant for computing the final target adjoint
+            values. Default is False.
+        """
+        # Work out other cases when they arise.
+        assert last_block == 0
```

Comment:
Should probably `raise NotImplementedError` or something. Assertion errors are just for internal code.

Reply:
done.